docs(contributing): intrsuction for developing framework-specific styles

asudoh opened the pull request:
#### Changelog

**New**

- Guide for developing (exceptional) JavaScript framework-specific styles in `carbon-components` repo.

asudoh committed:
docs(contributing): intrsuction for developing framework-specific styles

joshblack commented:
Just for context, who/what is this for?

asudoh commented:
@joshblack I have seen some custom styles in Angular, etc. repos. Most of them should be eliminated or kept in the framework variant repos, but there may be some cases where adding styles to this repo makes sense. That said, this is for providing the most efficient way for developers of such framework variant repos to fiddle with our styles.

joshblack commented:
In this case, would it be more appropriate to call out how to use `npm link` or `yarn link` similar to: https://github.com/IBM/carbon-components-react/blob/master/docs/carbon-components.md ?

asudoh commented:
@joshblack Thank you for the link. My instruction attempts to make the workflow more efficient, by removing the need of `gulp watch` (and thus the time overhead associated with that task). Please note that my `ln -s` instruction creates a symbolic link to `/path/to/fork/of/carbon-components/src` directory instead of `/path/to/fork/of/carbon-components/scss` directory. That said, `yarn link` usage is not optimal here. I'll try updating your link, too.

joshblack commented:
@asudoh not sure what you mean by optimal, it seems like a reasonable setup when running `carbon-components` alongside a framework implementation. No need to worry about symlinks and can use the tooling that we rely on or that we ourselves use to develop the project (like the npm scripts).

asudoh commented:
@joshblack The approach with `npm link` requires running `gulp watch` in `carbon-components` side, which involves a step of copying `.scss` files in `carbon-components/src` over to `carbon-componetns/scss`. The instruction in this PR attempts to eliminate it, though you are right the set-up and tooling is easier in your way.

tw15egan committed:
Merge branch 'master' into doc-framework-specific-style

tw15egan committed:
Merge branch 'master' into doc-framework-specific-style

joshblack commented:
I think using the development API, through `yarn dev`, could help protect against unexpected file structure changes. Not to say that they would be common, just that using how we develop could provide more reliable/future-proof instructions.

Let me know if I'm misunderstanding the use-case here, though! I might be missing something big here 😅

asudoh committed:
docs(contributing): add yarn link approach

asudoh commented:
Just in case I was not clear enough, I was talking about the the spent on the build step in `carbon-components` side, associated with `yarn gulp watch` or `yarn dev` - Approximately 3 seconds upon every `.scss` file change. Usage of `yarn dev` adds around 20s for initial start up.

Your point about reliability/future-proof-ness is fair, though. I have added `yarn link` approach as the primary way, and added a caveat of reliability/future-proof-ness to `ln -s` approach.

tw15egan committed:
Merge branch 'master' into doc-framework-specific-style

jnm2377 commented on .github/CONTRIBUTING.md:
```diff
@@ -141,6 +141,71 @@ What `.hbs` file is used for rendering a variant is determined by searching for
 
+## Working on JavaScript-framework-specific styles
+
+JavaScript-framework-specific is _not_ recommended as we strive to create styles that is framework-neutral. However, there are some rare cases where framework-specific cannot be avoided, and some of those make sense to be in maintained by core style library here.
```
This should be "styles that _are_ framework-neutral."

jnm2377 commented on .github/CONTRIBUTING.md:
```diff
@@ -141,6 +141,71 @@ What `.hbs` file is used for rendering a variant is determined by searching for
+
+### Pointing NPM dependency of `carbon-components` right to the source code
+
+Though above approach is the most straightforward, it involves an overhead of having to run build process at `cabron-components`, in addition to one at framework variant repo, upon every Sass code change.
```
Should be `carbon-components` here in the inline code.

jnm2377 reviewed (approved):
Looks good other then some small comments I left!

asudoh committed:
docs(contributing): apply review comments

asudoh commented:
Applied the suggested changes - Thanks @tw15egan for reviewing and @jnm2377 for your feedback!

asudoh committed:
Merge branch 'master' into doc-framework-specific-style
